msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/DeviceCodeFlowApiTest.java

```diff
@@ -0,0 +1,329 @@
+        final AuthorizationResult authorizationResult = mStrategy.getDeviceCode(authorizationRequest, mUrlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals(ErrorStrings.INVALID_REQUEST, authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureNoScope() throws IOException {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = mBuilder.setScope(null).build();
+        final AuthorizationResult authorizationResult = mStrategy.getDeviceCode(authorizationRequest, mUrlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals(ErrorStrings.INVALID_REQUEST, authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureBadScope() throws IOException {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = mBuilder.setScope("/").build();
+        final AuthorizationResult authorizationResult = mStrategy.getDeviceCode(authorizationRequest, mUrlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals(ErrorStrings.INVALID_SCOPE, authorizationErrorResponse.getError());
+    }
+
+    //===========================================================================================================
+    // Token Request Testing
+    //===========================================================================================================x
+    @Test
+    public void testDeviceCodeFlowTokenInvalidRequest() throws IOException, ClientException {
+        final MicrosoftStsOAuth2Configuration config = new MicrosoftStsOAuth2Configuration();
+        config.setAuthorityUrl(new URL(mUrlBody));
+        final OAuth2StrategyParameters options = new OAuth2StrategyParameters();
+        final OAuth2Strategy strategy = new MicrosoftStsOAuth2Strategy(config, options);
+
+        final TokenResult tokenResult = strategy.requestToken(mTokenRequest);
```

Comment:
nit: Let's not use a class variable. Can we create a function that returns a new (mock) token request instead.

Reply:
Will do, cleaned out most of the setup methods and got rid of most class variables.